### PR TITLE
[leaf_function] Fill None gradients with zeros in backward

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #178444
* __->__ #178422

In invoke_leaf_function's backward, autograd_grad_with_gradient_info can
return None for non-requires_grad tensor inputs. Replace None grads with
torch.zeros(...) to prevent crashes in downstream autograd when these
None values flow into ops that expect tensors.

Also fix fake_backward to produce zero gradients for tensor inputs
without requires_grad, which previously returned None.

### CODE REVIEW

File: torch/_higher_order_ops/invoke_leaf_function.py [hunk 1]
```diff
@@ -663,12 +663,23 @@ def real_backward(*grads):
...
                 output_infos=real_state["outputs"],
                 input_infos=real_state["inputs"],
                 grad_outputs=grads,
                 allow_unused=True,
             )
+            result = tuple(
+                g
+                if g is not None
+                else (
+                    torch.zeros(arg.size(), dtype=arg.dtype, device=arg.device)
```
Comment (aorenste) [inline torch/_higher_order_ops/invoke_leaf_function.py:676]: `torch.zeros_like(arg)`?

File: torch/_higher_order_ops/invoke_leaf_function.py [hunk 2]
```diff
@@ -683,15 +694,30 @@ def real_backward(*grads):
             for arg in flat_args
         )
 
+        # input_infos_for_fake is None for tensor inputs without requires_grad.
```
Comment (aorenste) [inline torch/_higher_order_ops/invoke_leaf_function.py:697]: We don't use input_infos_for_fake for anything other than creating the tensor below - maybe instead encode everything in input_infos_for_fake (like `(GradientInfo, arg.requires_grad)`) instead of having a separate tuple for the non-requires_grad case.
